### ISSUE TITLE
cmd/govim: create temporary files within a govim-specific $TMP sub-directory

### ISSUE THREAD

**Issue by gkotian** [govim/govim#1059]
### What version of Vim/Gvim are you using?

<details><summary>Vim/Gvim version</summary><br><pre>
$ vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Feb 09 2021 23:51:55)
Included patches: 1-2489
Compiled by Arch Linux
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             +mouse_sgr         +tag_binary
+arabic            +file_in_path      -mouse_sysmouse    -tag_old_static
+autocmd           +find_in_path      +mouse_urxvt       -tag_any_white
+autochdir         +float             +mouse_xterm       +tcl/dyn
-autoservername    +folding           +multi_byte        +termguicolors
-balloon_eval      -footer            +multi_lang        +terminal
+balloon_eval_term +fork()            -mzscheme          +terminfo
-browse            +gettext           +netbeans_intg     +termresponse
++builtin_terms    -hangul_input      +num64             +textobjects
+byte_offset       +iconv             +packages          +textprop
+channel           +insert_expand     +path_extra        +timers
+cindent           +ipv6              +perl/dyn          +title
-clientserver      +job               +persistent_undo   -toolbar
-clipboard         +jumplist          +popupwin          +user_commands
+cmdline_compl     +keymap            +postscript        +vartabs
+cmdline_hist      +lambda            +printer           +vertsplit
+cmdline_info      +langmap           +profile           +virtualedit
+comments          +libcall           +python/dyn        +visual
+conceal           +linebreak         +python3/dyn       +visualextra
+cryptv            +lispindent        +quickfix          +viminfo
+cscope            +listcmds          +reltime           +vreplace
+cursorbind        +localmap          +rightleft         +wildignore
+cursorshape       +lua/dyn           +ruby/dyn          +wildmenu
+dialog_con        +menu              +scrollbind        +windows
+diff              +mksession         +signs             +writebackup
+digraphs          +modify_fname      +smartindent       -X11
-dnd               +mouse             -sound             -xfontset
-ebcdic            -mouseshape        +spell             -xim
+emacs_tags        +mouse_dec         +startuptime       -xpm
+eval              +mouse_gpm         +statusline        -xsmp
+ex_extra          -mouse_jsbterm     -sun_workshop      -xterm_clipboard
+extra_search      +mouse_netterm     +syntax            -xterm_save
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 
Linking: gcc -L. -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.32/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -Wl,--as-needed -o vim -lm -ltinfo -lelf -lacl -lattr -lgpm -ldl -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.32/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -L/usr/local/lib -L/usr/lib/perl5/5.32/core_perl/CORE -lperl -lpthread -ldl -lm -lcrypt -lutil -lc -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lm
</pre></details>

### What version of Go are you using (`go version`)?

<details><summary>Go version</summary><br><pre>
$ go version
go version go1.16 linux/amd64
</pre></details>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/gautam/.cache/go-build"
GOENV="/home/gautam/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/gautam/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/gautam/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2665515259=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What version/commit of `govim` are you using?
a8b8ff21e5a8e67a4f251e7d4d6d67c7a49b5a43

### What is the output of `:messages` in Vim?
nothing afaict.

### Do the Vim channel, `govim` or `gopls` log files show anything interesting?

I haven't looked into any of them.

### How did you install `govim`?

vundle

### What did you do?

just used it :)

### What did you expect to see?
My `/tmp` directory remaining uncluttered.<br>
A few files generated by the plugin would be acceptable, a single directory created by the plugin (with all its temporary files placed inside) would be ideal.

### What did you see instead?
A whole bunch of `govim*`, `gopls*` & `vim_channel*` files.

**Comment by myitcv** [comment]
👍  from me. I suffer this problem too. The suggestion is clean and simple and I like it.

@leitzler any thoughts? Can you see anything to suggest this would be a bad idea?

**Comment by leitzler** [comment]
Sounds like a good idea!

We do write both logs and socket files for parent-child communication. I think this change should focus on logs only and keep the sockets in tmp. In other words, the `/tmp` folder might still end up with some `govim-parent-child-*` directories.

Reason being that sockets doesn't feel "temporary" in the same sense as logs, you can delete the logs at any time but a socket might still be in use, right?

**Comment by myitcv** [comment]
Excellent point about the sockets. How about this:

* "namespace" create temporary log files under `$TMP/govim`, as opposed to just `$TMP`
* create socket files as `$TMP/.govim-parent-child-*`, i.e. hidden

Socket files are generally of far less (no?) interest to the user. Making them "hidden" helps reflect that. But not bundling them with the logs reflects your very good point, @leitzler, that they are a different beast to log files. 

Thoughts?

**Comment by gkotian** [comment]
>sockets doesn't feel "temporary" in the same sense as logs

Good point. But then is `/tmp` really the right location for them? Would a govim-specific directory in `/run` or `/var/run` be a better location for the socket files?
But I'm not sure how this impacts support for other platforms/other distributions. (I use Arch Linux where `/var/run` is just a symlink to `/run`).

**Comment by gkotian** [comment]
Alternatively, how about:
- `$TMP/govim/logs` for log files
- `$TMP/govim/sockets/govim-parent-child-*` for socket files

That said, keeping log files in a different directory and making the socket directories hidden would already solve my problem, so that also works for me.

**Comment by myitcv** [comment]
> But then is `/tmp` really the right location for them? Would a govim-specific directory in `/run` or `/var/run` be a better location for the socket files?

Honestly I have no idea :) The only requirement is that it be user-writable and secure from a multi-user system perspective. 


> Alternatively, how about:

That would work, but it would be relatively easy to blow away `$TMP/govim` without thinking about it. And that would kill all parent-child sockets in use.

I'll ask around for ideas on this. But for how, the hidden socket approach seems the "best" of a bad bunch!

**Comment by gkotian** [comment]
>relatively easy to blow away

One option would be to make it `$TMP/.govim` so it would be a bit harder to inadvertently delete.

If you want precedent, I see `/tmp/.org.chromium.Chromium.*` on my system which I suspect is slack.

**Comment by gkotian** [comment]
According to https://unix.stackexchange.com/a/162911/13642, perhaps `/run/user/$UID/govim` is what you're looking for :)

**Comment by myitcv** [comment]
> perhaps `/run/user/$UID/govim` is what you're looking for

That does indeed seem to tick the relevant boxes.

**Comment by leitzler** [comment]
I agree that `/run/user/$UID/govim` sounds like the best solution, long term. 

I do however think that we should keep it somewhere in `/tmp` as long as we can't "guarantee" that we do not leave directories and/or sockets behind, as per: https://github.com/govim/govim/blob/3f63a3d43ebb2dc92940c3ed59df155ad3c6a55d/cmd/govim/main.go#L370-L371

As a loose reference:
```
$ uptime
 18:41:46 up 8 days,  7:20,  7 users,  load average: 0,00, 0,00, 0,00
 $ find /tmp -name "govim-parent-child-*" | wc -l
70
```
and
```
$ find /tmp/govim-parent* -name "socket" | wc -l
28
```

While I only have 1 govim session alive.

**Comment by myitcv** [comment]
Wow. That is bad.

**Comment by myitcv** [comment]
@leitzler what we could do in the meantime is handle process signals within the `govim` process to do that tidy up.

**Comment by myitcv** [comment]
But of course the ultimate fix here is to, per that linked comment, fix the shutdown protocol.

**Comment by myitcv** [comment]
Ok, here is a consolidated suggestion on next steps. Please up/down vote, critique etc:

* move socket file creation to subdirectories within `/run/user/$UID/govim`. This suggestion seems to be backed up by others in `#tools` on Slack which is good
* (we could consider switching to using `net.FileListener` and `net.FileConn` so that we only need to create temporary files, but that's a smaller detail)
* handle process signals to ensure that on `SIGINT` (assuming that's what we are getting, we should do some analysis to confirm) we tidy up our temporary socket files
* move logging to files within `$TMP/govim`
* work on fixing https://github.com/govim/govim/issues/842 if we get time (at which point the signal handling should be used as a panic assertion mechanism in case we haven't got our shutdown sequence right)

Thoughts?

**Comment by leitzler** [comment]
Another aspect in favour of keeping socket files under $TMP is that `/run` doesn't exist in macOS. It does have `/var/run` but that seems to be for system level sockets and doesn't include any `user` directory.

The other bullets sounds like good (independent) ideas!

**Comment by myitcv** [comment]
> Another aspect in favour of keeping socket files under $TMP is that `/run` doesn't exist in macOS. It does have `/var/run` but that seems to be for system level sockets and doesn't include any `user` directory.

`gopls` uses [`$XDG_RUNTIME_DIR`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables) if set but this does not really help us. 

So I think, given @leitzler's observation, we need to stick with `$TMP`. I've amended the proposal above.

**Comment by myitcv** [comment]
Turns out there were broader problems associated with us not tidying up properly; namely that `gopls` temporary files would also be left lying around. Hence I've just pushed up #1062 to fix everything in one go (the changes make sense to combine to my mind, because we can then be clear that we are properly shutting down, so that `govim` and `gopls` can tidy up).

**Comment by myitcv** [comment]
Picking up on @leitzler's excellent observation in https://github.com/govim/govim/pull/1062#discussion_r600004153, we need to revise the decision of a `$TMP/govim` root directory for log files - it doesn't work on shared user systems.

`$TMP/govim_$UID` feels awkward, which suggests we reconsider the approach of hidden log files, e.g. `$TMP/.govim-*.log`.

Incidentally, the fixes in #1062 are working perfectly on my local machine. No more socket directories/files left lying around (not `gopls` files for that matter). So it's now down to the naming of the log files, because I think hidden socket directories + files "works".

**Comment by myitcv** [comment]
@leitzler has once again come up with a better suggestion: namely that we allow people to opt out of log files being created. This is also a step towards disabling log file creation by default.

So the new suggestion is that we use the value of the `GOVIM_LOG` environment variable to determine whether or not to create log files. 

See https://github.com/govim/govim/pull/1064
